torch/csrc/jit/import_source.cpp

```diff
@@ -416,58 +430,79 @@ struct SourceImporter {
     return size_t(version.asIntegral());
   }
 
-  void parseImportsAndDoCallback() {
-    // Gather all imports
-    auto& L = p_.lexer();
-    std::vector<std::string> imports;
+  // older versions of serialization required import statements,
+  // and defined classes file-at-a-time in import order.
+  // The problem is that in Python
+  // it is possible to construct cyclic dependencies between files even
+  // when there are none between individual classes. New versions of loading
+  // just compile class-at-a-time, so we no longer need to follow the import
+  // order. Future serialization may stop producing the import code.
+  void parseImports(Lexer& L) {
     while (L.nextIf(TK_IMPORT)) {
       std::ostringstream s;
       while (L.cur().kind != TK_NEWLINE) {
         s << L.cur().text();
         L.next();
       }
       L.expect(TK_NEWLINE);
-      const auto str = s.str();
-      AT_ASSERT(!str.empty());
-      imports.push_back(str);
-    }
-
-    // Call theregister_typectually compile them
-    for (const auto& import : imports) {
-      if (import_callback_) {
-        import_callback_(import);
-      }
     }
   }
 
-  Parser p_;
-  size_t version_;
   std::shared_ptr<CompilationUnit> cu_;
-  const std::function<void(const std::string&)>& import_callback_;
-  const std::vector<at::Tensor>& tensor_table_;
-  std::shared_ptr<SourceResolver> resolver_;
+  std::unordered_map<std::string, std::shared_ptr<SugaredValue>> env_;
+  SourceLoader source_loader_;
+  std::unordered_set<std::string> loaded_sources_;
+  // named types and functions loaded from a file but not yet defined because
+  // their type has not been requested yet.
+  std::unordered_map<std::string, TreeRef> to_be_defined_;
```

Comment:
do we define hash for QualifiedName? If so, we can make this a map of QualifiedName directly instead of `std::string`. I think it would make the code clearer

Reply:
Looks like we do, so I can change this.